lib/index.js

```diff
@@ -79,17 +79,32 @@ PostHTML.prototype.process = function (tree, options) {
    * @prop {?Boolean} options.sync - enables sync mode, plugins will run synchronously, throws an error when used with async plugins
    * @prop {?Function} options.parser - use custom parser, replaces default (posthtml-parser)
    * @prop {?Function} options.render - use custom render, replaces default (posthtml-render)
-   * @prop {?Boolean} options.skipParse - disable parsing
    */
   options = options || {}
 
-  if (options.parser) parser = options.parser
-  if (options.render) render = options.render
+  if (typeof tree !== 'string' || typeof tree !== 'object') {
+    throw new TypeError(
+      'Input must to be a String (HTML) or Object (PostHTML AST)'
+    )
+  }
+
+  tree = typeof tree === 'object' ? tree : parser(tree)
+
+  tree.parser = parser
+  tree.render = render
 
-  tree = options.skipParse ? tree : parser(tree)
+  if (options.parser && typeof options.parser === 'function') {
+    parser = options.parser
+  }
+  if (options.render && typeof options.render === 'function') {
+    render = options.render
+  }
 
+  tree.plugins = this.plugins
   tree.options = options
 
+  tree.messages = []
```

Comment:
example 👍. `this.messages` + `PostHTML.prototype.messages`

```js
PostHTML.prototype.messages = function (message) {
   this.messages = []

   if (message)  {
     this.messages.push(message)
   }

   return this
}
```

```js
posthtml.messages({...locals})
```

?

Currently its in `PostHTML.prototype.process` like options, messages is 'per process'. 
Should all properties be thisified then? e.g `this.options` and why ?